Switch to Swagger for api documentation

caspermeijn opened the pull request:
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | How to start?
| Documentation | no
| Translation   | no
| CHANGELOG.md  | Do I just start a new section?
| License       | MIT

<!--
Please list the issues your PR fixes using special keywords, see
https://help.github.com/articles/closing-issues-using-keywords/

Fixes #…
-->

<!--
- Please fill in this template according to the PR you're about to submit.
- Replace this comment by a description of what your PR is solving.
-->

I work on [Read It Later](https://gitlab.gnome.org/World/read-it-later) wallabag client and I want better api documentation. I think the first step is updating nelmio/api-doc. This will bring an extended way to document the api.

This PR updates to nelmio/api-doc 3.0 and translates the existing documentation to the new format. I tried to keep the same content as the old api doc.

I have two outstanding issues:
- The `{_format)` path parameters are not automatically shown in version 3.0. I don't see how this was added in the old version. Which api calls do have the format parameter? I assume this is only a subset of the calls, so want to add these manually.
- nelmio/api-doc  4.0 is not possible, because is will result in a conflict. Do you see a simple way to resolve that conflict?

Makes progress on https://github.com/wallabag/wallabag/issues/2721

j0k3r reviewed (commented):
Good job!

To run test: `make test`
Also, run CS-Fixer too: `php bin/php-cs-fixer fix`

caspermeijn commented on src/Wallabag/CoreBundle/Command/SwaggerDocblockConvertCommand.php:
```diff
@@ -0,0 +1,226 @@
+        if ($apiDoc->getResource()) {
+            throw new \RuntimeException('implement me');
+        }
+        $path = str_replace('.{_format}', '', $apiDoc->getRoute()->getPath());
```
No, I will remove the convert command code as that was a one time action.

I meant the old api docs have a _format in all calls. See this picture. I will assume only the export api call uses the {_format}.
![Screenshot _format](https://user-images.githubusercontent.com/1286074/200764848-bc2d35d1-862e-4943-a10b-8a56b84e650c.png)

j0k3r replied:
Oh yes, you might be right

caspermeijn commented:
I fixed the tests and ran the formatter. I added the _format parameter to the export call, however in this version of nelmio/api-doc the {_format} is always stripped of the path. Please see if this is acceptable.

j0k3r commented:
> however in this version of nelmio/api-doc the {_format} is always stripped of the path

You mean, on the generated doc?

caspermeijn commented:
> > however in this version of nelmio/api-doc the {_format} is always stripped of the path
> 
> You mean, on the generated doc?

Yes, in the generated doc the {_format} is always stripped for the path. That is hard coded in the nelmio code.

j0k3r commented:
I've pushed some doc update and added authorization stuff.

Here are some screenshot, before:
<img width="1298" alt="image" src="https://user-images.githubusercontent.com/62333/201928547-e64391e8-a0a9-447c-acac-79622c46046e.png">

After:
<img width="1298" alt="image" src="https://user-images.githubusercontent.com/62333/201928677-445e2052-0f9c-4a2e-90a2-45e35364dce4.png">

j0k3r commented:
> * nelmio/api-doc  4.0 is not possible, because is will result in a conflict. Do you see a simple way to resolve that conflict?

What was that conflict?

Kdecherf reviewed (approved):
LGTM

caspermeijn commented:
> > * nelmio/api-doc  4.0 is not possible, because is will result in a conflict. Do you see a simple way to resolve that conflict?
> 
> What was that conflict?

`composer update` generates a lot of output. But I think this is main problem:
```
- nelmio/api-doc-bundle[v4.0.0-BETA1, ..., v4.0.1] require symfony/property-info ^4.0|^5.0 -> satisfiable by symfony/property-info[v4.0.0-BETA1, ..., 4.4.x-dev, v5.0.0-BETA1, ..., 5.4.x-dev].
```

caspermeijn commented:
> Can you squash your commits? And we'll merge it. Thanks for the work!

Done
